8346420: C2: IfNode::fold_compares_helper() wrongly folds two CmpI nodes to a single CmpU node

eme64 opened the pull request:
**Introduction**

The optimization in `IfNode::fold_compares_helper` tries to optimize:
- 2 `CmpI` -> constant fold the second
- 2 `CmpI` -> turn into one `CmpU`

There has been a history of bugs here, and I'm finally adding modular arithmetic proofs.
Making a small point-fix was not so easy, so I decided on a rewrite.
Added lots of testing.

----------------------

**The Reproducer(s)**

But a bug was discovered in one case, where we get an overflow:
```
    @Test
    @Arguments(values = {Argument.NUMBER_42})
    // Reported overflow case with wrong result in JDK-8346420
    public static void test_Case3a_LTLE_overflow(int i) {
        int minimum, maximum;
        if (FLAG_FALSE) {
            minimum = 0;
            maximum = 1;
        } else {
            // Always goes to else-path
            minimum = Integer.MIN_VALUE;
            maximum = Integer.MAX_VALUE;
        }
        // i  < INT_MIN    || i  > MAX_INT
        // 42 < INT_MIN    || 42 > MAX_INT
        //    false           false
        // => false
        //
        // C2 transforms this into:
        // i  - minimum >=u (maximum - minimum) + 1
        // 42 - INT_MIN >=u (INT_MAX - INT_MIN) + 1
        // 42 + MIN_INT >=u -1                  + 1
        //                  ------ overflow -------
        // 42 + MIN_INT >=u 0
        // => true
        if (i < minimum || i > maximum) {
            throw new RuntimeException("i can never be outside [min_int, max_int]");
        }
    }
```

We need to investigate the transformation logic, and look at the "explanations", and replace them with solid modular arithmetic proofs.

But once playing around with reproducers, I also found a case that produced a `fatal error: no reachable node should have no use`. We use a `hook` for `lo`, so it does not disappear. But in some cases, we actually have no use for the node any more, and we need to actively destroy it - I added logic for that now.

----------------------------------------

**History: No Proof = Hiding Bugs**

There was a similar bug in this code before, fixed by [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823), [see Roland's PR email](https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2015-June/018181.html). And another one [JDK-8135069](https://bugs.openjdk.org/browse/JDK-8135069).

These were the comments / "explanations" he added. I find them a bit hard to read, so I'll annotate them. Sadly, they are no proofs, and so underflow/overflow can easily slip through:
```
    // this_bool = <
    //   dom_bool = >= (proj = True) or dom_bool = < (proj = False)
    //     x in [a, b[ on the fail (= True) projection, b > a-1 (because of hi_type->_lo > lo_type->_hi test above):
    //     lo = a, hi = b, adjusted_lim = b-a, cond = <u
```
- In`x >= a && x < b` or `!(x < a) && x < b`
- Out `x - lo <u adjusted_lim` = `x - a <u b-a`
- corresponds to my case `c)` - my proofs show this is correct


```
    //   dom_bool = > (proj = True) or dom_bool = <= (proj = False)
    //     x in ]a, b[ on the fail (= True) projection, b > a:
    //     lo = a+1, hi = b, adjusted_lim = b-a-1, cond = <u
```
- In `x > a && x < b` or `!(x <= a) && x < b`
- Out `x - lo <u adjusted_lim` = `x - (a+1) <u b-(a+1)`
- corresponds to my case `a)` - my proofs show this is correct



```
    // this_bool = <=
    //   dom_bool = >= (proj = True) or dom_bool = < (proj = False)
    //     x in [a, b] on the fail (= True) projection, b+1 > a-1:
    //     lo = a, hi = b, adjusted_lim = b-a+1, cond = <u
    //     lo = a, hi = b, adjusted_lim = b-a, cond = <=u doesn't work because b = a - 1 is possible, then b-a = -1
```
- In `x >= a && x <= b` or `!(x < a) || x <= b`
- Out pre [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823): 
  - `x - lo <u adjusted_lim` = `x - a <=u b-a`
  - Bug if `b = a - 1`, e.g. `a = 0` and `b = -1`: `x >= 0 && x <= -1` is always false, but `x - 0 <=u -1 = max_unit` is always true, the rhs underflows!
- Out past [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823): 
  - `x - lo <u adjusted_lim` = `x - a <u b-a + 1`
  - Bug if `b = a + 1`, i.e. if `a=min_int` and `b=max_int`: `x >= min_int && x <= max_int` always true but `x - min_int <u max_int-min_int+1 = 2^32 = 0` is aways false, the rhs overflows!
- corresponds to my case `d)` - this is the buggy case. [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823) just converted an underflow but into an overflow bug. There was no proper proof - and so we did not catch this.


```
    //   dom_bool = > (proj = True) or dom_bool = <= (proj = False)
    //     x in ]a, b] on the fail (= True) projection b+1 > a:
    //     lo = a+1, hi = b, adjusted_lim = b-a, cond = <u
    //     lo = a+1, hi = b, adjusted_lim = b-a-1, cond = <=u doesn't work because a = b is possible, then b-a-1 = -1
```
- We convert `x > a && x <= b` or `!(x <= a) && x <= b`
- Out pre [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823):
  - `x - lo <u adjusted_lim` = `x - (a+1) <=u b-a-1`
  - Bug if `a=b`, e.g. `a=b=0`: `x > 0 && x <= 0` always false, but `x - (0+1) <=u 0-0-1 = max_uint` is always true, because of overflow!
- Out post [JDK-8081823](https://bugs.openjdk.org/browse/JDK-8081823):
  - `x - lo <u adjusted_lim` = `x - (a+1) <u b-a`
- corresponds to my caee `b)` - my proofs show this is correct

-----------------------------

**Solution**

I tried for a long time to keep the same code structure, and keep the fix more minimal.
But it got harder and harder. There are not just 4 cases, but all sorts of permutations.

Some of the "bug-cases" of case `d)` would have needed to instead go to the `2 CmpI -> constant fold the second one` optimization (`x >= 0 && x <= -1` can be constant folded, no need for `CmpU`), but the cases were in separate if-else blocks. So a bit of a mess.

The naming also kept on confusing me. How upper bounds become lower bounds, and success condition and fail conditions are (mis)named.

So I decided I was just going to rewrite `fold_compares_helper` from the top. This allows me to name things consistently and restructure the optimizations.

I'm still a bit frustrated by the naming in the surrounding code. The uncommon trap (often path to out-of-bounds-exception) are called "success", while the in-bounds path is called "fail". But I won't rename everything now in this bug-fix PR.

**Testing**

I have some IR tests that cover a lot of cases, and some "fuzzing" style template test as well.

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8346420](https://bugs.openjdk.org/browse/JDK-8346420): C2: IfNode::fold_compares_helper() wrongly folds two CmpI nodes to a single CmpU node (**Bug** - P3)


### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - **Reviewer**)
 * [Galder Zamarreño](https://openjdk.org/census#galder) (@galderz - Committer) Review applies to [0f34fd28](https://git.openjdk.org/jdk/pull/28819/files/0f34fd28ed965e50ac2d8e09c9173794a9b2690b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28819/head:pull/28819` \
`$ git checkout pull/28819`

Update a local copy of the PR: \
`$ git checkout pull/28819` \
`$ git pull https://git.openjdk.org/jdk.git pull/28819/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28819`

View PR using the GUI difftool: \
`$ git pr show -t 28819`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28819.diff">https://git.openjdk.org/jdk/pull/28819.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28819#issuecomment-4268989802)
</details>

eme64 committed:
JDK-8346420

eme64 committed:
more cases

eme64 committed:
constant fold cases

eme64 committed:
minor update

eme64 committed:
array length and null check

eme64 committed:
wip fuzzer

eme64 committed:
brainstorming

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

eme64 committed:
wip comparator for fuzzer

eme64 committed:
some random test code

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

eme64 committed:
refactor a bit

eme64 committed:
add some more cases to test

eme64 committed:
refactor flags in test

eme64 committed:
fix warmup and comments in tests

eme64 committed:
uncomment failing tests

eme64 committed:
Working on fix

eme64 committed:
fix up documentation and assert in IfNode::filtered_int_type

eme64 committed:
cleanup a bit

eme64 committed:
fix assert issue with dead node

eme64 committed:
more cleanup

eme64 committed:
yet more cleanup

eme64 committed:
more comments and updates

eme64 committed:
fix up comments

eme64 committed:
add some delimiter lines

eme64 committed:
more clarifications

eme64 committed:
more comments

eme64 committed:
rm old explanations

eme64 committed:
cleanup and add assert

eme64 committed:
fixing up notation wip

eme64 committed:
more renaming / cleanup

eme64 committed:
more corrections

eme64 committed:
more comments

eme64 committed:
wip fix

eme64 committed:
some more details

eme64 committed:
the tricky case

eme64 committed:
wip move around proofs

eme64 committed:
fix case b

eme64 committed:
fix case c and some other details

eme64 committed:
another small fix

eme64 committed:
manual merge

eme64 committed:
implement fix

eme64 committed:
fix small issue

eme64 committed:
fix up test and add one more comment

eme64 committed:
result verification for one method

eme64 committed:
add randomized IR test

eme64 committed:
more extreme cases

eme64 committed:
wip IR rule generation

eme64 committed:
small test refactor and fix

eme64 committed:
make rules precise for one case

eme64 committed:
wip IR rules

eme64 committed:
fix yet another case, and refactor

eme64 committed:
wip next stuff

eme64 committed:
more progress on IR rules

eme64 committed:
generate and/or cases

eme64 committed:
random inputs

eme64 committed:
more comments

eme64 committed:
more comments

eme64 committed:
wire in warmup to IR rules

eme64 committed:
fix IR rule under warmup

eme64 committed:
fuzzing switch case

eme64 committed:
wip arr length test$

eme64 committed:
cleanup fuzzer test

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

eme64 commented:
/template append

eme64 committed:
fix IR rule for case with warmup

rwestrel commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+  //     Using a >= min_int, and b <= max_int:
+  //     a + 2^32 <= min_int + 2^32
+  //               = max_int + 1
+  //              <= b       + 1
```
Should it be `>=` here and `>` below?

eme64 replied:
Right, all the inequalities are flipped!

eme64 replied:
Another typo it is.

rwestrel commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+    //
+    //   Case lo < n < hi:
+    //     (BEFORE) is always true, show (AFTER) is always true.
+    //     Since lo < hi (LO-HI), S(lo+1) = lo+1 (no overfaw):
```
typo: overflow

rwestrel commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+    // QED.
+    //
+    // Note: we cannot use anything more relaxed than the assumption
+    //       lo < hi: with lo=hi the rhs of the CmpU would undeflow.
```
typo: underflow

rwestrel reviewed (approved):
Other than the typos, looks good to me.

rwestrel commented:
>     * 2 `CmpU` -> turn into one `CmpU`

Does it really do that?
AFAICT, it only attempts a transformation if it sees a `CmpI`.

eme64 commented:
> > ```
> > * 2 `CmpU` -> turn into one `CmpU`
> > ```
> 
> Does it really do that? AFAICT, it only attempts a transformation if it sees a `CmpI`.

Thanks for catching it! That was a typo, sorry :see_no_evil:

eme64 commented:
@rwestrel Thanks for catching the typos! I fixed them now. And thanks for the approval :)

Ready for a second reviewer!

eme64 committed:
Roland's suggestions

galderz commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -663,49 +669,66 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
         const CmpNode* cmp  = bol->in(1)->as_Cmp();
+        // Val is always the lhs of the comparision: val <test> cmp2
         if (cmp->in(1) == val) {
+          assert(cmp->Opcode() == Op_CmpI, "signed comparison required");
```
I don't know how far fetched this is, but one of the callers of `IfNode::filtered_int_type` is `CountedLoopConverter::filtered_type_from_dominators`. Could there be a loop situation where you arrive here with a `CmpU`? I don't see any loop tests in `TestFoldCompares` that can validate this and from what I can decipher from the fuzzer, I don't see any there either. E.g.

```
for (i = 0; i < ...; i++) {
  if () { // TODO a condition that converts to a CmpU
    arr[i] = ...;
  }
}
```

eme64 replied:
Maybe, I would have to check. Not sure if `CountedLoopConverter::filtered_type_from_dominators` allows for `CmpU`.

No testing failed for this assert, but maybe we never generate such patterns.
You are right, we could generate some examples with the fuzzer, just in case.

eme64 replied:
Added unsigned comparisons to the fuzzer. I think that would shake out examples that should hit the assert, if the method is ever used in the wrong context.

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -663,49 +669,66 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
             case BoolTest::eq:
+              // Condition leading to if_proj: val == cmp2
+              //   val in cmp2_t
               return cmp2_t;
```
Should this be `val_t->join(cmp2_t)`?

eme64 replied:
Maybe... but I'm not sure of the consequences. It could be that we get similar issues as discussed below.

Would you be ok leaving this as is for now?

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -663,49 +669,66 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
             case BoolTest::lt:
+              // Condition leading to if_proj: val < cmp2
+              //   val in [min_int .. max(min_int, cmp2->_hi - 1)]
               lo = TypeInt::INT->_lo;
```
Should just use `min_jint` here.

eme64 replied:
done

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -663,49 +669,66 @@ const TypeInt* IfNode::filtered_int_type(PhaseGVN* gvn, Node* val, Node* if_proj
+              return nullptr;
             }
             const TypeInt* rtn_t = TypeInt::make(lo, hi, cmp2_t->_widen);
             return rtn_t;
```
Should this be joined with `val_t`?

eme64 replied:
I think that would have bad consequences, and would break some optimizations.

Example:
```
// i in [-100 .. 100]
if (i < 0 || i > 10) {
    failure
}
```

Let's look at the check `i < 0`. The failure path value `rtn_t` would be `[min_int .. -1]`.
And so currently, in the optimization, we check that the two checks section the 3 paths into 3 regions. The lo failure path must thus include `min_int`, and the hi failure path must include `max_int`.

If we filter with the type of `i`, then we get `[-100 .. 100]` joinded with `[min_int .. -1]` -> `[-100 .. -1]`, and that does not include `min_int`. And the optimization fails.

merykitty replied:
Hmmm, what I am confused about is that in one of the cases (ne), we take into consideration `val_t`, while in the others, we only look at the type infered by the comparison. Why is there such an inconsistency, and what this function is supposed to return?

eme64 replied:
That is a good question. The restriction seems unnecessary, and would prevent optimizations down the line. We could consider removing it in a follow-up RFE, and ensure that one side of the returned interval touches `min_int` or `max_int`.

I'm afraid I might break something in this change if I do it now, so I'd like to avoid an enhancement while implementing a fix. Is that ok?

> what this function is supposed to return?

I think the issue is that this is not very precisely defined / stated.

My understanding: it has to provide a type that is a subtype of the range that the comparison constrains `val` to under the branch. But also a subtype of this constraint intersected with the type of `val` before the branch.

Optimizations afterwards will assume that `val` will only have values in that type under this branch, which allows us to section the whole int-range into disjunct regions - this is required for the CmpI/CmpU transform.

eme64 replied:
Right now, the `IfNode::fold_compares_helper` expects that the `lo_type` and `hi_type` touch the `min_int` and `max_int`, respectively. But really what we need: three disjunct regions - `lo_type`, `middle_type` and `hi_type`, that are ordered from small to large. If we adjusted the optimization like that, we could then allow the types not to touch `min_int` or `max_int`, and we could let `IfNode::filtered_int_type` return the intersection of the comparison-constraint and the prior `val` type. I think that would be more elegant and general.

But I don't want to do that now in this fix ;)

eme64 replied:
If you agree with my assessment, we could improve the code comments, and file a follow-up RFE to improve the code further - and make sure no invariants are broken for any possible other use of `IfNode::filtered_int_type`.

merykitty replied:
That makes sense, thanks for answering my question.

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+  IfNode* iff2 = this;
+  BoolNode* bool1 = iff1->in(1)->as_Bool();
+  BoolNode* bool2 = iff2->in(1)->as_Bool();
+  CmpNode* cmp1 = bool1->in(1)->as_Cmp();
```
Assert that these are `CmpINode`s.

eme64 replied:
done

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+  if (type_middle != nullptr) {
+    const TypeInt* type_succ = filtered_int_type(igvn, n, succ);
+    if (type_succ != nullptr) {
+      if (type_middle->filter(type_succ) == Type::TOP) {
```
Do we try to work the case when `fail2` is unreachable?

eme64 replied:
I suppose not here. But there could well be other optimizations that do it. We can make it a future investigation, if you want. What do you think?

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+      if (type_middle->filter(type_succ) == Type::TOP) {
+        // The intersection is empty -> succ is not reachable.
+        // Fold iff2 towards fail2 (and away from succ).
+        igvn->replace_input_of(iff2, 1, igvn->intcon(fail2->_con));
```
When you remove an `IfNode`, you need to rewire all nodes below it that `depends_only_on_test`.

eme64 replied:
I'm not removing the `IfNode`, only hacking its condition code. Folding itself happens later during IGVN, right?

Or what exactly are you suggesting? We also had this optimization before, and did no rewiring of the nodes...

merykitty replied:
Hacking the condition IS folding the node. If you look at loop predication, we will also hack the conditional code and let IGVN do the actual folding. There, we call `PhaseIdealLoop::dominated_by`, in which the conditional code is hacked, then we call `PhaseIdealLoop::rewire_safe_outputs_to_dominator` to move `depends_only_on_test` nodes to the correct control.

merykitty replied:
Logically, after you hack the conditional code, you do not know why this `If` is a constant anymore, and you cannot rewire its outputs correspondingly. So, the only logical place to do that is when you hack the conditional code where you know exactly why it can be hacked.

eme64 replied:
I still don't understand. Can you please give me some more explanation?

We just figure out that one path is never taken, and so we replace the condition with a constant. Is that so different to when a If condition constant folds during IGVN? We don't do the rewiring there either, right?

The optimization here is part of `IfNode::Idea` -> `IfNode::fold_compares`. So we could not use `PhaseIdealLoop::rewire_safe_outputs_to_dominator`.

Loop predication seems like a bit of a different case, no? Because there, we actually move a check out of the loop, and so rewiring has to happen from within the loop to outside the loop. But here, the dependencies would just float to before the if that is being constant folded, and that would still be at an ok place, no?

eme64 replied:
After discussing offline with @merykitty and @chhagedorn , we came to this conclusion:
- The issue would be pre-existing, and we can investigate it separately, and fix it separately.
- The concern is that some node is only dependent on `iff1`.
  - If there is no null-check, then it would be ok, since it would now be automatically attached to `iff2`, which is a sufficient condition.
  - But if there is a `null-check` between, then it would now be automatically attached to that, and if the null-check gets hoisted, it would be hoisted over `iff2`, and that is not ok. We should probably just pin the node anyway (just to be safe), but it would be interesting to see if such a case is even possible, and find a reproducer.

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
+    //
+    // Note: we cannot use anything more relaxed than the assumption
+    //       lo < hi: with lo=hi the rhs of the CmpU would underflow.
+    if (lo_type->_hi >= hi_type->_lo) {
```
We should move this precondition to before the proof, because the `assuming lo < hi` is declared there.

eme64 replied:
Sure, I can move them up, I'll do that for all 4 cases.

eme64 replied:
done

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -898,219 +921,509 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
-  igvn->replace_input_of(this, 1, newbool);
+  // Fold iff1 towards middle, and replace the iff2 condition:
+  igvn->replace_input_of(iff1, 1, igvn->intcon(middle->_con));
+  igvn->replace_input_of(iff2, 1, newbool);
```
Similarly, you need to rewire the nodes below each `If`s that `depends_only_on_test`.

eme64 replied:
Same response: I'm only hacking the condition codes. And we did not do anything like that before. Can you explain a bit more?

merykitty reviewed (commented):
I have some comments regarding the VM change.

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

eme64 committed:
add unsigned comparisons to fuzzer test

eme64 committed:
cleanup

eme64 commented:
@merykitty @galderz I finally got the time to come back to this. I've answered / implemented all of your points, I think. Can you please have another look?

eme64 committed:
review suggestions by Quan Anh

eme64 commented:
@galderz I now found a fuzzer failure with `Integer.compareUnsigned`. Not sure if it is related or another bug. Investigating.

But thanks already for the idea, with testing unsigned comparisions! :heart:

eme64 commented:
Ok, I found the issue. In `RegionNode::optimize_trichotomy`, we can fold two consecutive comparisons. We do a similar trick as in the fold-compares optimization.

But it seems we don't check if the two `Cmp` nodes have the same opcode ... so now we fold `CmpI` with `CmpU`, and that's not correct.

```
// java -XX:CompileCommand=compileonly,Test4::test* -XX:CompileCommand=printcompilation,Test4::test* -Xcomp Test4.java

public class Test4 {
    public static void main(String[] args) {
        boolean gold = testA1(256);

        for (int i = 0; i < 10000; i++) {
            boolean val = testA1(256);
            if (val != gold) { throw new RuntimeException("wrong value: " + gold + " vs " + val); }
        }
    }

    public static boolean flag = true;

    static boolean testA1(int n) {
        int a = flag ? -3 : Integer.MAX_VALUE;
        // RegionNode::optimize_trichotomy
        // thinks that the comparisons below operate on the same inputs compare(n, a),
        // but misses to check the Opcode for CmpI vs CmpU.
        if (a > n || Integer.compareUnsigned(n, a) > 0) {
            return true;
        }
        return false;
    }
}
```

That's really a separate bug though.
But now my fuzzer finds these cases, so we can't integrate it together. Maybe I'll have to revert the unsigned-comparison fuzzer additions for now.

Update: filed bug: https://bugs.openjdk.org/browse/JDK-8385157

eme64 commented:
@merykitty @galderz I now removed the unsigned comparisons from the fuzzer, because of [JDK-8385157](https://bugs.openjdk.org/browse/JDK-8385157). I'll keep investigating the `depends_only_on_test` issue, but if I find something, I'll file it as a separate bug anyway.

I think I addressed all concerns here. Could you please re-review?

eme64 committed:
remove unsigned comparisons from fuzzer again because of JDK-8385157

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

eme64 commented:
@merykitty @galderz I could not find any failing case for the `depends_only_on_test` issue. I think it is currently impossible, though I can't prove it. We could consider a conservative pinning-fix in a follow-up RFE, just to be safe.

But what about the changes here, are they ok, and ready to integrate?

merykitty commented:
I think it is ok then, there is still a point I am not sure about:

> Hmmm, what I am confused about is that in one of the cases (ne), we take into consideration val_t, while in the others, we only look at the type infered by the comparison. Why is there such an inconsistency, and what this function is supposed to return?

eme64 commented:
@merykitty thanks for bringing that back up. I responded up here: https://github.com/openjdk/jdk/pull/28819#discussion_r3308799582

eme64 committed:
Merge branch 'master' into JDK-8346420-CmpI-CmpU-fix

merykitty commented on src/hotspot/share/opto/ifnode.cpp:
```diff
@@ -902,219 +925,523 @@ bool IfNode::has_only_uncommon_traps(IfProjNode* proj, IfProjNode*& success, IfP
 
-  PhaseTransform::SpeculativeProgressGuard progress_guard(igvn);
-  // convert:
+  // Optimization 2: try to replace the two CmpI with one CmpU
```
Can you add a comment here explaining how the transformation is legal, as I have told you in DM. This comment does not really explain how the graph is transformed and how the nodes should be rewired as a result of that transformation.

eme64 replied:
There are already comments about the rewiring elsewhere, so I'm not sure if we should repeat that here:
https://github.com/eme64/jdk/blob/0f34fd28ed965e50ac2d8e09c9173794a9b2690b/src/hotspot/share/opto/ifnode.cpp#L743
https://github.com/eme64/jdk/blob/0f34fd28ed965e50ac2d8e09c9173794a9b2690b/src/hotspot/share/opto/ifnode.cpp#L928

What exactly more do you need me to add here?

merykitty replied:
I feel that those descriptions are inadequate and they do not really explain what happens here. For example, if we see the transformation as:

    if (x >= 0 && x < n) into if (x u< n)

Then it must be the case that there is nothing between `if (x >= 0)` and `if (x < n)`, which is definitely not the case when we want to do the transformation when there are null checks in between. I think decomposing the transformation into:

    if (x >= 0) {
        null-checks;
        if (x < n) {
            succ();
        }
    }
    fail();

    1. Clone the whole graph through `if (x u< n)`:

    if (x u< n) {
        if (x >= 0) {
            null-checks;
            if (x < n) {
                succ();
            }
        }
        fail();
    } else {
        if (x >= 0) {
            null-checks;
            if (x < n) {
                succ();
            }
        }
        fail();
    }

    2. Elide a test because there is already a dominating test that implies it:

    if (x u< n) {
        null-checks;
        succ();
    } else {
        if (x >= 0) {
            null-checks;
        }
        fail();
    }

Then it is clearer why the transformation is legal, it also makes it clear where the depending nodes should go, and the null checks need cloning in the `if (x u>= n)` branch.

eme64 replied:
Right, so I think we agree that if there is no null-check, the transformation is correct, and no further documentation would be needed.

You are now asking about code outside the bug I'm fixing, which is only about the modular arithmetic issues in `IfNode::fold_compares_helper`.

To answer your question, we'd have to look at `IfNode::fold_compares`.
It alows these patterns:
- `is_ctrl_folds`: nothing between the `iff1` and `iff2`. So the transformation is correct, as agreed on.
  - `IfNode::has_shared_region`. The failure path go to the same region.
  - `IfNode::has_only_uncommon_traps`. The failure paths both lead to traps.
- Else: a check could be in between. So now we do extra checks:
  - `has_only_uncommon_traps`: both failure paths of `iff1` and `iff2` must go to traps.
  - `is_side_effect_free_test`: that "other" check must also be side-effect free and failure of it also leads to a trap.
  - This has the following effect: all checks are side-effect free, and so we can execute them more freely, and if any of them fails, we can just trap, and let the interpreter sort out the exact behavior.
  - You may also want to read `reroute_side_effect_free_unc`, that explains that we use the trap of `iff1`.

So I think it is all there, and there are explanations. Though really not much at `IfNode::fold_compares` that puts it all together. Also: the naming there is quite bad, and I have only "fixed" the naming inside `IfNode::fold_compares_helper`.

I'm considering refactoring/documenting more of this code in a follow-up RFE. Would that be good enough, or do you think it is best that I do all the renaming and documentation with the bug-fix?

merykitty reviewed (approved):
Looks good to me, just a small request, I think it also relates to the fact that you want to add TODO regarding `depends_only_on_test` nodes.

galderz reviewed (approved):
Thanks for addressing concerns with unsigned comparisons. Happy with the follow up

eme64 commented:
@merykitty Thanks for all the input. I now filed a follow-up RFE for the issues we had discussed:
https://bugs.openjdk.org/browse/JDK-8385570
- More documentation/proofs/renaming.
- `depends_only_on_test` and pinning in null-check case.
- `IfNode::filtered_int_type`: consistency/guarantees/assumptions - and maybe optimization improvements.

@galderz Thanks for re-reviewing!

eme64 commented:
Ah, and of course we found this bug, thanks to @galderz 's suggestion :)
https://bugs.openjdk.org/browse/JDK-8385157

eme64 commented:
I just encountered a small test-bug in the fuzzer IR rules. I fixed it and am running testing again now.

eme64 committed:
small fix/relaxation for IR rules in warmup-case of fuzzer

eme64 commented:
@galderz @merykitty @rwestrel Testing is running, and completed successfully everywhere except `macosx-aarch64` :snail: 

I'll need some re-approval, I hope to integrate on Monday :)

eme64 commented:
@galderz @merykitty @rwestrel Thank you for the many good suggestions and comments, and for the approval!

I'll let you know how it goes with the follow-up and related bug.
https://bugs.openjdk.org/browse/JDK-8385570
https://bugs.openjdk.org/browse/JDK-8385157

/integrate
